src/util/strencodings.h

```diff
@@ -369,4 +371,78 @@ std::string Capitalize(std::string str);
  */
 std::optional<uint64_t> ParseByteUnits(std::string_view str, ByteUnit default_multiplier);
 
+namespace util {
+/** consteval version of HexDigit() without the lookup table. */
+consteval uint8_t ConstevalHexDigit(const char c)
+{
+    if (c >= '0' && c <= '9') return c - '0';
+    if (c >= 'a' && c <= 'f') return c - 'a' + 0xa;
+
+    throw "Only lowercase hex digits are allowed, for consistency";
+}
+
+/**
+ * ""_hex is a compile-time user-defined literal returning a
+ * `std::array<std::byte>`, equivalent to ParseHex(). Variants provided:
+ *
+ * - ""_hex_v: Returns `std::vector<std::byte>`, useful for heap allocation or
+ *   variable-length serialization.
+ *
+ * - ""_hex_u8: Returns `std::array<uint8_t>`, for cases where `std::byte` is
+ *   incompatible.
+ *
+ * - ""_hex_v_u8: Returns `std::vector<uint8_t>`, combining heap allocation with
+ *   `uint8_t`.
+ *
+ * @warning It may be necessary to use _v variants when serializing, or vice
+ *   versa, because vectors are assumed to be variable-length and serialized
+ *   with a size prefix, while arrays are considered fixed length and serialized
+ *   with no prefix.
+ *
+ * @warning It may be preferable to use _v variants to save stack space when
+ *   declaring local variables if hex strings are large. Alternately variables
+ *   could be declared constexpr to avoid using stack space.
+ *
+ * @warning Avoid _u8 variants when not necessary, as the codebase migrates to
+ *   use `std::byte` instead of `unsigned char` and `uint8_t`.
+ *
+ * @note The reason ""_hex uses `std::array` is because a heap-based
+ *   `std::vector` cannot cross the compile-time/runtime barrier.
```

Comment:
In commit "util: Add consteval ""_hex[_v][_u8] literals" (f9fed4d4e6daa2bdd715b92b6bfffdbe8202a9c8)

Maybe change "The reason" to "One reason". I think even if C++ did provide a way to convert `constexpr` vectors to runtime vectors(*) it would still make sense for this to use std::array because:

- Semantically this returns a fixed size array, not a variable sized one
- std::array is more flexible because array values can be allocated on the heap, or on stack, or in static storage, while std::vector is inflexible and can basically only allocate values on a heap
- Array type is simpler and doesn't add unnecessary overhead.

(*) C++ doesn't provide a way to convert constexpr vectors to runtime vectors, but it could, and I assume we could even write our own `ToRuntime<constexpr_vector>()` function doing this conversion.

Reply:
Yeah, my goal was to contrast it against `ParseHex` returning a vector. Will do if I re-touch.